core/src/impl/Kokkos_GraphImpl_fwd.hpp

```diff
@@ -0,0 +1,60 @@
+/*
+//@HEADER
+// ************************************************************************
+//
+//                        Kokkos v. 3.0
+//       Copyright (2020) National Technology & Engineering
+//               Solutions of Sandia, LLC (NTESS).
+//
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_KOKKOS_GRAPHIMPL_FWD_HPP
+#define KOKKOS_IMPL_KOKKOS_GRAPHIMPL_FWD_HPP
+
+#include <Kokkos_Macros.hpp>
+
+namespace Kokkos {
+namespace Impl {
+
+struct KernelInGraphProperty {};
+
+struct IsGraphKernelTag {};
```

Comment:
Any reason why you prefer having two tag classes here.  In principle we could use just one.

Reply:
One would be fine probably. I think when I implemented this I was thinking that there may be different requirements associated with tag classes and properties at some point in the future, but yeah, of course they could be the same thing for now. I also think I just thought I'd be less likely to make mistakes if they were separate things, but I'm not opinionated either way.

Reply:
My (weak) preference goes to a single tag.

Reply:
I think my (weak) preference is to leave it as is because of the expected infrastructure changes coming down the line surrounding properties. But if it's going to hold up the pull request, I'll change it. I certainly don't care that much.

Reply:
If we were to remove the public overloads of require with an impl argument we could drop one of the tags I guess?

Reply:
This would make it so that the changes we already know we want to make to `require` are not backwards compatible. I still don't think this is worth changing here. Are either of you holding up the pull request based on this, or can we move on? (Minor rant: if you leave a comment and don't say anything, you're by default saying that you want to keep the pull request from merging until this is addressed, so if it's not your intention to hold up this whole pull request on this basis we should be clearer about that...)